qutebrowser/app.py

```diff
@@ -814,6 +826,22 @@ def on_focus_object_changed(self, obj):
             log.misc.debug("Focus object changed: {}".format(output))
         self._last_focus_object = output
 
+    @pyqtSlot(Qt.ApplicationState)
+    def on_app_state_changed(self, state):
```

Comment:
I'm not sure if we really have to do this. The [Qt docs](http://doc.qt.io/qt-5/qguiapplication.html#applicationState) say:

> You can react to application state changes to perform actions such as stopping/resuming CPU-intensive tasks, freeing/loading resources or saving/restoring application data.

And [for `ApplicationActive`](http://doc.qt.io/qt-5/qt.html#ApplicationState-enum):

> The application is visible and selected to be in front.

So isn't this just Qt telling us that the window is raised? Do we really have to do anything here, or does Qt just raise it by itself?

Reply:
This method is what raises QB window after normal click on a link in the other app. At first I used `ApplicationActivate` event, but looks like it's the same. This func is being called in two cases (that I'm aware of):

* User switched to the app (e.g. on ⌘-Tab)
* We got an "activate" event from apple

In the first case there is nothing to be done, indeed.

Reply:
But doesn't Qt already handle that event by itself? That's what I get from reading the documentation.

Reply:
I don't know about Qt in general, but QB does not raise on link being clicked if this event is not handled explicitly.

Reply:
I'll need to look into this, but I still feel like it'd be best to remove it entirely. FWIW, it breaks things when starting qutebrowser for me:

```
Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/qutebrowser/utils/objreg.py", line 300, in last_focused_window
    return get('last-focused-main-window')
  File "/home/florian/proj/qutebrowser/git/qutebrowser/utils/objreg.py", line 229, in get
    return reg[name]
  File "/home/florian/proj/qutebrowser/git/.venv/lib/python3.6/collections/__init__.py", line 993, in __getitem__
    raise KeyError(key)
KeyError: 'last-focused-main-window'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/qutebrowser/app.py", line 834, in on_app_state_changed
    window = objreg.last_focused_window()
  File "/home/florian/proj/qutebrowser/git/qutebrowser/utils/objreg.py", line 302, in last_focused_window
    return window_by_index(-1)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/utils/objreg.py", line 308, in window_by_index
    raise NoWindow()
qutebrowser.utils.objreg.NoWindow
```

Reply:
I guess it's on Linux. Seems like the event arrives before the first window is created. Hmm...

Reply:
There is one more bug regarding this code. When app is being activated via click on one of it's windows, the last focused window is being raised instead of the one user clicked.

Reply:
Let's drop it then. We can still re-visit this after merging the PR, but at the moment this seems to introduce more problems than it solves, and I'd like to merge this PR before hopefully releasing v1.0.0 later this week.